blog/06-02-2023-catering-to-the-client-side.md

```diff
@@ -0,0 +1,173 @@
+  client->>+provider: evaluateFlags(evaluationContext)
+  provider-->>-client: flag values
+  client->>+cache: store(flag values)
+  cache->>-client: .
+  %% deactivate cache
+  client->>-app: .
+  %% deactivate client
+  note right of app: some time later..
+  app->>+client: operation that needs a flagging decision
+  client->>+cache: getFlagValue(flagKey)
+  note right of cache: no call to flagging service needed
+  cache->>-client: previously evaluated value
+  client->>-app: .
+
+
+```
+
+Put another way, we seperate flag [**evaluation**](/docs/specification/glossary#evaluating-flag-values) - passing an evaluation context through a set of rules in order to determine a flagging decision - from flag [**resolution**](/docs/specification/glossary#resolving-flag-values) - getting the flagging decision for a specific feature flag.
+
+## Client-side support in OpenFeature
+
+With OpenFeature we have been thinking about how to support these key differences between client-side and server-side feature flagging. We think about these differences as two paradigms: dynamic context (for server-side flags) and static context (for client-side flags). 
+
+OpenFeature's current Evaluation API supports the dynamic paradigm quite nicely, but to support the static paradigm (and thus client-side flagging) we need to add a second flavor of the Evaluation API. 
+
+### Server-side evaluation today
+Let's compare and contrast. A typical server-side flagging decision using OpenFeature's current SDK might look something like this:
+``` java
+@GetMapping("/hello")
+public String getSalutation() {
+    final Client client = openFeatureAPI.getClient();
+    final evalContext:EvaluationContext = evalContextForCurentRequest();
+
+    if (client.getBooleanValue("use-formal-salutation", false, evalContext)) {
+      return "Good day to you!";
+    }else{
+      return "Hey, what's up?";
+    }
+} 
+```
+You can see that we're passing evaluation context every time we need to make a flagging decision.
+
+### Client-side evaluation tomorrow
+With the currently proposed OpenFeature changes, a client-side flagging decision would look more like this:
+
+``` java
+public string generateSalutation(){
+  if (client.getBooleanValue("use-formal-salutation", false)) {
+    return "Good day to you!";
+  }else{
+    return "Hey, what's up?";
+  }
+}
+```
+
+You can see we are no longer passing any evaluation context when requesting a flagging decisions. 
+
+But our flagging framework needs to get flag evaluation context from somewhere. So rather than passing the context as part of each flagging decision, the app provides a new evaluation context to OpenFeature whenever a user logs in or out (as part of this app's hypothetical `onAuthenticated(...)` handler): 
+
+``` java
+// called whenever a user logs in (or out)
+public void onAuthenticated(userId:String){
+  OpenFeatureAPI api = OpenFeatureAPI.getInstance();
+  api.setEvaluationContext(new MutableContext().add("targettingKey", userId));
```

Comment:
Perhaps we should use JavaScript in the example since it may be easier for a front-end dev to provide feedback. This example is pretty straightforward though, so it's completely up to you.

Reply:
I was reluctant to use JS for this first example because the reader might get distracted by the sync vs async difference in the two flavors of JS APIs. I also sort of think it would be nice to have something other than JS, for readers who are more Java than JavaScript.

Reply:
Yep, that makes sense.